lact-gui/src/app/graphs_window/plot/render_thread.rs

```diff
@@ -0,0 +1,362 @@
+use super::cubic_spline::cubic_spline_interpolation;
+use super::to_texture_ext::ToTextureExt;
+use super::PlotData;
+use anyhow::Context;
+use cairo::{Context as CairoContext, ImageSurface};
+
+use gtk::gdk::MemoryTexture;
+use itertools::Itertools;
+use plotters::prelude::*;
+use plotters::style::colors::full_palette::DEEPORANGE_100;
+use plotters_cairo::CairoBackend;
+use std::cmp::max;
+use std::ops::{Deref, DerefMut};
+use std::sync::{Arc, Mutex};
+use thread_priority::{ThreadBuilderExt, ThreadPriority};
+use tracing::error;
+
+enum Request {
+    Terminate,
+    Render(RenderRequest),
+}
+
+#[derive(Default)]
+pub struct RenderRequest {
+    pub title: String,
+    pub value_suffix: String,
+    pub secondary_value_suffix: String,
+    pub y_label_area_size: u32,
+    pub secondary_y_label_area_size: u32,
+
+    pub data: PlotData,
+
+    pub width: u32,
+    pub height: u32,
+
+    pub supersample_factor: u32,
+}
+
+#[derive(Default)]
+struct RenderThreadState {
+    request_condition_variable: std::sync::Condvar,
+    last_texture: Mutex<Option<MemoryTexture>>,
+    current_request: Mutex<Option<Request>>,
+}
+
+/// A rendering thread that will listen for rendering requests and process them asynchronously.
+/// Requests that weren't processed in time or resulted in error are dropped.
+pub struct RenderThread {
+    /// Shared state is between the main thread and the rendering thread.
+    state: Arc<RenderThreadState>,
+    thread_handle: Option<std::thread::JoinHandle<()>>,
+}
+
+/// Ensure the rendering thread is terminated properly when the RenderThread object is dropped.
+/// We send Request::Terminate to swiftly terminate rendering thread and then join to let it finish last render.
+impl Drop for RenderThread {
+    fn drop(&mut self) {
+        self.state
+            .current_request
+            .lock()
+            .unwrap()
+            .replace(Request::Terminate);
+        self.state.request_condition_variable.notify_all();
+
+        self.thread_handle.take().map(|handle| handle.join().ok());
+    }
+}
+
+impl RenderThread {
+    pub fn new() -> Self {
+        let state = Arc::new(RenderThreadState::default());
+
+        let thread_handle = std::thread::Builder::new()
+            .name("Plot-Renderer".to_owned())
+            // Render thread is very unimportant, skipping frames and rendering slowly is ok
+            .spawn_with_priority(ThreadPriority::Min, {
```

Comment:
Just curious - what does the thread priority do on a system level? Does it set the `nice` value?

Previously I've used glib's `idle_add` for background work, but that runs on the main thread so it's not a good fit here.

Reply:
I think issue lies with my usage, as I like to fullscreen all the nice graphs we have.

And detaching it from GUI thread gives us room for higher supersampling. As currently this is 4x faster, because we render in parallel.

Reply:
I think this is just a good practice to set explicit priority for this threads.

It didn't make difference for me personally, as I have high thread count CPU and almost never have CPU starvation for all of them. 

LACT before this PR was using 10% of single core for me (which is quite significant, considering I have AMD Ryzen 9 7950x CPU)

The idea is that graphs are not doing vital work and we are okay to run them at low frame rate. I wouldn't want all the fancy rendering to interfere much with someone's benchmarks.

As you understand, this is all educated guess work, but don't think it should have same priority as GUI thread.

Reply:
It calls this syscall https://man7.org/linux/man-pages/man2/sched_setattr.2.html

Reply:
> LACT before this PR was using 10% of single core for me (which is quite significant, considering I have AMD Ryzen 9 7950x CPU)

This is weird, because on my slower 3700x it's <2% in a release build. But still it makes sense to draw this at a lower refresh rate, I agree.